File: src/main/kotlin/com/cp/csms/transactions/controllers/TransactionController.java
package com.cp.csms.transactions.controllers;

import com.cp.csms.common.AuthenticationStatus;
import com.cp.csms.transactions.AuthorizationRequest;
import com.cp.csms.transactions.AuthorizationResponse;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.concurrent.TimeoutException;


@Slf4j
@RestController
@RequiredArgsConstructor
@RequestMapping(value = "api/v1/transaction")
public class TransactionController {

    private final AuthorizationService authorizationService;

    @PostMapping("/authorize")
    public ResponseEntity<AuthorizationResponse> authorizeTransaction(@RequestBody AuthorizationRequest request) {
        log.info("Authorizing transaction for request: {}", request);

        try {
            final AuthenticationStatus authenticationStatus = authorizationService.authorize(request);
            log.info("Authorization result for request {}: {}", request, authenticationStatus);
            return ResponseEntity.ok(
                    AuthorizationResponse.builder()
                            .authenticationStatus(authenticationStatus)
                            .build()
            );
        } catch (TimeoutException e) {
            return ResponseEntity.status(HttpStatus.GATEWAY_TIMEOUT).build();
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

}

File: src/main/kotlin/com/cp/csms/transactions/controllers/TransactionController.kt
package com.cp.csms.transactions.controllers

import com.cp.csms.transactions.AuthorizationRequest
import com.cp.csms.transactions.AuthorizationResponse
import com.cp.csms.transactions.AuthorizationService
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.concurrent.TimeoutException

@RestController
@RequestMapping("api/v1/transaction")
class TransactionController(
    private val authorizationService: AuthorizationService
) {

    @PostMapping("/authorize")
    fun authorizeTransaction(
        @RequestBody request: AuthorizationRequest
    ): ResponseEntity<AuthorizationResponse> {
        log.info("Authorizing transaction for request: {}", request)

        return try {
            val authenticationStatus = authorizationService.authorize(request)
            log.info("Authorization result for request {}: {}", request, authenticationStatus)
            ResponseEntity.ok(AuthorizationResponse(authenticationStatus))
        } catch (e: TimeoutException) {
            ResponseEntity.status(HttpStatus.GATEWAY_TIMEOUT).build()
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build()
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(TransactionController::class.java)
    }
}
